crates/bevy_pbr/src/light_probe/environment_filter.wgsl

```diff
@@ -0,0 +1,427 @@
+        case 56u: { dir = vec3f(0.213508, 0.976688, 0.0222359); }
+        case 57u: { dir = vec3f(0.536034, -0.10141, -0.838084); }
+        case 58u: { dir = vec3f(-0.147707, -0.941925, 0.301597); }
+        case 59u: { dir = vec3f(-0.822975, 0.102675, -0.558722); }
+        case 60u: { dir = vec3f(0.0753368, 0.810439, 0.580958); }
+        case 61u: { dir = vec3f(0.958193, -0.0618399, -0.279361); }
+        case 62u: { dir = vec3f(-0.0168296, -0.509477, 0.860319); }
+        case 63u: { dir = vec3f(-0.999999, 0.00159255, 0.0); }
+        default: { dir = vec3f(0.0, 0.0, 1.0); }
+    }
+    
+    return normalize(dir);
+}
+
+// Convert UV and face index to direction vector
+fn sample_cube_dir(uv: vec2f, face: u32) -> vec3f {
+    // Convert from [0,1] to [-1,1]
+    let uvc = 2.0 * uv - 1.0;
+    
+    // Generate direction based on the cube face
+    var dir: vec3f;
+    switch(face) {
+        case 0u: { dir = vec3f( 1.0,  -uvc.y, -uvc.x); } // +X
+        case 1u: { dir = vec3f(-1.0,  -uvc.y,  uvc.x); } // -X
+        case 2u: { dir = vec3f( uvc.x,  1.0,   uvc.y); } // +Y
+        case 3u: { dir = vec3f( uvc.x, -1.0,  -uvc.y); } // -Y
+        case 4u: { dir = vec3f( uvc.x, -uvc.y,  1.0);  } // +Z
+        case 5u: { dir = vec3f(-uvc.x, -uvc.y, -1.0);  } // -Z
+        default: { dir = vec3f(0.0); }
+    }
+    return normalize(dir);
+}
+
+// Convert direction vector to cube face UV
+struct CubeUV {
+    uv: vec2f,
+    face: u32,
+}
+fn dir_to_cube_uv(dir: vec3f) -> CubeUV {
+    let abs_dir = abs(dir);
+    var face: u32 = 0u;
+    var uv: vec2f = vec2f(0.0);
+    
+    // Find the dominant axis to determine face
+    if (abs_dir.x >= abs_dir.y && abs_dir.x >= abs_dir.z) {
+        // X axis is dominant
+        if (dir.x > 0.0) {
+            face = 0u; // +X
+            uv = vec2f(-dir.z, -dir.y) / dir.x;
+        } else {
+            face = 1u; // -X
+            uv = vec2f(dir.z, -dir.y) / abs_dir.x;
+        }
+    } else if (abs_dir.y >= abs_dir.x && abs_dir.y >= abs_dir.z) {
+        // Y axis is dominant
+        if (dir.y > 0.0) {
+            face = 2u; // +Y
+            uv = vec2f(dir.x, dir.z) / dir.y;
+        } else {
+            face = 3u; // -Y
+            uv = vec2f(dir.x, -dir.z) / abs_dir.y;
+        }
+    } else {
+        // Z axis is dominant
+        if (dir.z > 0.0) {
+            face = 4u; // +Z
+            uv = vec2f(dir.x, -dir.y) / dir.z;
+        } else {
+            face = 5u; // -Z
+            uv = vec2f(-dir.x, -dir.y) / abs_dir.z;
+        }
+    }
+    
+    // Convert from [-1,1] to [0,1]
+    return CubeUV(uv * 0.5 + 0.5, face);
+}
+
+// Sample an environment map with a specific LOD
+fn sample_environment(dir: vec3f, level: f32) -> vec4f {
+    let cube_uv = dir_to_cube_uv(dir);
+    return textureSampleLevel(input_texture, input_sampler, cube_uv.uv, cube_uv.face, level);
+}
+
+// Calculate tangent space for the given normal
+fn calculate_tangent_frame(normal: vec3f) -> mat3x3f {
```

Comment:
Assuming this is what I think it is, I suggest using https://github.com/JMS55/bevy/blob/solari6/crates/bevy_solari/src/scene/sampling.wgsl#L89-L97.

If you want to be cool you can move it into some shared file I can later use in bevy_solari :)

Reply:
Shared the function  called `build_orthonormal_basis` at bevy_pbr utils module